algorithms/src/snark/varuna/ahp/indexer/constraint_system.rs

```diff
@@ -46,20 +47,20 @@ impl<F: Field> ConstraintSystem<F> {
 
     #[inline]
     /// Returns the sparse A matrix as Vec of rows, where each row is a Vec of assigned value and variable index
-    pub(crate) fn a_matrix(&self) -> Result<Vec<Vec<(F, usize)>>> {
-        to_matrix_helper(&self.a, self.num_public_variables)
+    pub(crate) fn a_matrix(&mut self) -> Result<Vec<Vec<(F, usize)>>> {
+        to_matrix_helper(mem::take(&mut self.a), self.num_public_variables)
     }
 
     #[inline]
     /// Returns the sparse B matrix as Vec of rows, where each row is a Vec of assigned value and variable index
-    pub(crate) fn b_matrix(&self) -> Result<Vec<Vec<(F, usize)>>> {
-        to_matrix_helper(&self.b, self.num_public_variables)
+    pub(crate) fn b_matrix(&mut self) -> Result<Vec<Vec<(F, usize)>>> {
+        to_matrix_helper(mem::take(&mut self.b), self.num_public_variables)
     }
 
     #[inline]
     /// Returns the sparse C matrix as Vec of rows, where each row is a Vec of assigned value and variable index
-    pub(crate) fn c_matrix(&self) -> Result<Vec<Vec<(F, usize)>>> {
-        to_matrix_helper(&self.c, self.num_public_variables)
+    pub(crate) fn c_matrix(&mut self) -> Result<Vec<Vec<(F, usize)>>> {
+        to_matrix_helper(mem::take(&mut self.c), self.num_public_variables)
```

Comment:
This should be reverted, because it is actually **taking** the items now.

While in theory this is "safe" from usage, it is not the desired behavior of this interface.

Reply:
I see that after we call `ics.{a,b,c}_matrix()`, we still need to extract a few more members. Can you follow the same approach as we use for the prover? https://github.com/AleoHQ/snarkVM/blob/testnet3/algorithms/src/snark/varuna/ahp/prover/round_functions/mod.rs#L101